doc/decisions/0_drafts/metakey_semantics.md

```diff
@@ -0,0 +1,191 @@
+  `keyMeta()` can just return `NULL` and `keyNew()` can either ignore metadata arguments or fail if metadata is given.
+
+## Considered Alternatives
+
+### Completely read-only
+
+One option to enforce all requirements is to just make any `Key` that is created with the `KEY_NS_META` namespace entirely read-only after `keyNew()`.
+
+Because the `Key` is entirely read-only, it's value cannot change.
+Therefore, we can find out, if it was copied from a given `spec:/` key by doing a pointer comparison:
+
+```c
+KeySet * specMeta = keyMeta (specKey);
+KeySet * otherMeta = keyMeta (otherMeta);
+
+for (elektraCursor it = 0; it < ksGetSize (otherMeta); it++)
+{
+   Key * metaKey = ksAtCursor (otherMeta, it);
+   if (metaKey == ksLookup (specMeta, metaKey, 0))
+   {
+      // copied from spec
+   }
+}
+```
+
+To update metadata you would have to create a new `Key` to replace the existing one:
+
+```c
+ksAppendKey (keyMeta (key), keyNew ("meta:/type", KEY_VALUE, "string", KEY_END));
+
+// would always fail, because metakeys are read-only
+keySetString (ksLookupByName (keyMeta (key), "meta:/type", 0), "string");
+```
+
+Safely copying metadata would be very simple with this solution.
+
+```c
+ksAppend (keyMeta (dest), keyMeta (source))
+```
+
+This copies metadata while sharing the memory for the individual `Key`s.
+Because the `Key`s are read-only they cannot be changed at all, so we don't have to worry about changes to `dest` affecting `source`.
+If we want to change the metadata of `dest` we have to create a new `Key`, which will not be used by `source`.
+
+### Read-only while in `KeySet`
+
+A relaxation of the above solution would be to only make the `Key` read-only, while it is part of a `KeySet`.
+
+The snippets from above would still work and changing the value of metakey directly as above would still fail.
+But now you don't have to create the `Key` directly with the right name and value.
+
+```c
+Key * metaKey = keyNew ("meta:/", KEY_END);
+keyAddBaseName (metaKey, somePart);
+keySetString (metaKey, someValue);
+```
+
+The snippet above would not work, if the key is returned as read-only directly from `keyNew()`.
+
+Copying metadata works the same as above.
+Since the `Key`s are still read-only as long as `source` uses them, we again cannot affect `source` by changing `dest`.
+
+### Utilize COW implementation
+
+An issue with the solutions above is that you always have to create a new `Key` to change metadata.
+This means we need to allocate all the memory for a new metakey.
+This new metakey will replace the existing one in `keyMeta(key)`.
+But that means, if the existing metakey was not shared with other keys, it will be deleted, when we could have just reused that memory.
+
+One option to solve that issue, in "Read-only while in `KeySet`" is something like this:
+
+```c
+Key * metaKey = ksLookupByName (keyMeta (key), "meta:/type", KDB_O_POP);
+if (isInKeySet (metaKey)) // <- hypothetical API
```

Comment:
huh, if you found the key with `ksLookupByName` then of course the key is in a keyset

Reply:
`ksLookup` with `KDB_O_POP` removes the key before it returns it. We want to remove that option, but we don't have any good replacement yet, so I used it for this example.

So the question here really is: Is this key in another keyset as well?